examples/src/main/java/kafka/examples/ExactlyOnceMessageProcessor.java

```diff
@@ -0,0 +1,190 @@
+        // across the consumer group instances.
+        if (this.mode.equals("groupMode")) {
+            consumer.subscribe(Collections.singleton(inputTopic), new ConsumerRebalanceListener() {
+                @Override
+                public void onPartitionsRevoked(Collection<TopicPartition> partitions) {
+                    printWithTxnId("Revoked partition assignment to kick-off rebalancing: " + partitions);
+                }
+
+                @Override
+                public void onPartitionsAssigned(Collection<TopicPartition> partitions) {
+                    printWithTxnId("Received partition assignment after rebalancing: " + partitions);
+                    messageRemaining.set(messagesRemaining(consumer));
+                }
+            });
+        } else {
+            // Do a range assignment of topic partitions.
+            List<TopicPartition> topicPartitions = new ArrayList<>();
+            int rangeSize = numPartitions / numInstances;
+            int startPartition = rangeSize * instanceIdx;
+            int endPartition = Math.min(numPartitions - 1, startPartition + rangeSize - 1);
+            for (int partition = startPartition; partition <= endPartition; partition++) {
+                topicPartitions.add(new TopicPartition(inputTopic, partition));
+            }
+
+            consumer.assign(topicPartitions);
+            printWithTxnId("Manually assign partitions: " + topicPartitions);
+        }
+
+        int messageProcessed = 0;
+        while (messageRemaining.get() > 0) {
+            ConsumerRecords<Integer, String> records = consumer.poll(Duration.ofMillis(200));
+            if (records.count() > 0) {
+                try {
+                    // Begin a new transaction session.
+                    producer.beginTransaction();
+                    for (ConsumerRecord<Integer, String> record : records) {
+                        // Process the record and send to downstream.
+                        ProducerRecord<Integer, String> customizedRecord = transform(record);
+                        producer.send(customizedRecord);
+                    }
+                    Map<TopicPartition, OffsetAndMetadata> positions = new HashMap<>();
+                    for (TopicPartition topicPartition : consumer.assignment()) {
+                        positions.put(topicPartition, new OffsetAndMetadata(consumer.position(topicPartition), null));
+                    }
+                    // Checkpoint the progress by sending offsets to group coordinator broker.
+                    // Under group mode, we must apply consumer group metadata for proper fencing.
+                    if (this.mode.equals("groupMode")) {
+                        producer.sendOffsetsToTransaction(positions, consumer.groupMetadata());
+                    } else {
+                        producer.sendOffsetsToTransaction(positions, consumerGroupId);
```

Comment:
Just to refresh my memory: are we going to eventually deprecate this API, or are we going to keep both, and let users apply this one with manual assignment (like you did here)? I thought we are going to deprecate, but maybe I remembered it wrong.

Reply:
No, we are intended to keep both